### ISSUE TITLE
Documentation Inconsistent Regarding .preventDefault() Events

### ISSUE THREAD

**Issue by pittins** [CreateJS/EaselJS#558]
tl:dr; Documentation on some events (such as EaselJS.Stage.drawstart) indicate you can cancel the drawing by calling `Event.preventDefault()`, but all Events are created as uncancelable events, making the documentation in error.

The actual quote on the EaselJS.Stage.drawstart event type is: "Dispatched each update immediately before the canvas is cleared and the display list is drawn to it. You can call preventDefault on the event object to cancel the draw." http://www.createjs.com/Docs/EaselJS/classes/Stage.html#event_drawstart

(Note that EaselJS.Stage.drawstart is used as an example, but this applies to others as well.)

(Also note, that I'm ignoring the work-around of just manually setting `evt.cancelable = true` in the listener itself: The raiser of the event should be able to decide with finality whether the event is cancelable or not.)

It looks like the Stage.update() code handles the case of a canceled drawstart event, but it doesn't actually create cancelable events. Now the easy thing to do here, is to just change how the event is created and dispatched:

line 374: http://www.createjs.com/Docs/EaselJS/files/easeljs_display_Stage.js.html#l363
```javascript
if (this.dispatchEvent("drawstart")) { return; }
```
would become
```javascript
if (this.dispatchEvent(new createjs.Event("drawstart", false, true))) { return; }
```

And in fact, doing so does seem to allow an event listener to call `evt.preventDefault()` on Event `evt`, and the drawing is in fact stopped. (And doing a similar thing for the tickstart event also appears to work just fine for that.)

I can't help but wonder though, if a possible change to EventDispatcher might be in order, to allow you to create events as `.dispatchEvent("drawstart", false, true)`.

I bring this up, because the current code in `.dispatchEvent()` optimizes away even the creation of an Event instance if it does not bubble and there are no listeners. Expanding upon this idea, an edited method could optimize away creating the event if it does not bubble and there are no listeners attached. Something like:

line 305: http://www.createjs.com/Docs/EaselJS/files/createjs_events_EventDispatcher.js.html#l287
```javascript
p.dispatchEvent = function(eventObj) {
  if (typeof eventObj == "string") {
    // won't bubble, so skip everything if there's no listeners:
    var listeners = this._listeners;
    if (!listeners || !listeners[eventObj]) { return false; }
    eventObj = new createjs.Event(eventObj);
  } ...
```
which would become
```javascript
p.dispatchEvent = function(eventObj, bubbles, cancelable) {
  if (typeof eventObj == "string") {
    // won't bubble, so skip everything if there's no listeners:
    if (!bubbles) {
      var listeners = this._listeners;
      if (!listeners || !listeners[eventObj]) { return false; }
    }
    eventObj = new createjs.Event(eventObj, !!bubbles, !!cancelable);
  } ...
```

The `bubbles/cancelable` parameters would just be ignored if `eventObj` is an actual Event object.

Slight performance decreases in having to perform more work in actually creating the new Event object, but slight performance increase in being able to optimize away creation of the Event object when there are no listeners (and a cancelable Event is desired).

Thoughts?

**Comment by MannyC** [comment]
I like your solution.  Note that prior to 0.8 `preventDefault` didn't actually check the `cancelable` property which would explain why it hasn't come up before ( see https://github.com/CreateJS/EaselJS/commit/b867154919a22138b6213a041b107c98b6dd3dad ), ie you could actually cancel uncancelable events.

**Comment by gskinner** [comment]
Good catch. I'll take a look at implementing this fix, or something similar.

**Comment by gskinner** [comment]
PS. if you have a list of other events you've noticed that suffer from this, please feel free to provide it here. We don't have many cancellable events, and they aren't the easiest things to track down via a straight find/replace.

**Comment by pittins** [comment]
Just did a search of all `dispatchEvent`, in EaselJS 0.8.0, PreloadJS 0.6.0, TweenJS 0.6.0, FlashAudioPlugin 0.6.0, MovieClip 0.8.0, SoundJS 0.6.0 (all combined varieties), and the only ones I've noticed are:

```
EaselJS.Stage.update() [drawstart]
EaselJS.Stage.tick() [tickstart]
```

Looks like nowhere else is there a dispatchEvent that is saved to a variable or used as a condition.

Hmm, also a tidbit from http://www.w3.org/TR/DOM-Level-2-Events/events.html#Events-EventTarget-dispatchEvent regarding the return value of dispatchEvent:

```
If preventDefault was called the value is false, else the value is true.
```

Haven't tested it, but it seems to indicate that `EaselJS.EventDispatcher.dispatchEvent()` is backwards, if you are looking at the DOM Level 2 Event Model. I'm not sure how much you are trying to match existing standards, but if it's intentional then an expanded comment in `EaselJS.EventDispatcher.dispatchEvent()` might be appropriate.

http://www.createjs.com/Docs/EaselJS/classes/EventDispatcher.html#method_dispatchEvent

**Comment by gskinner** [comment]
@pittins you're correct. I'm not even sure how that mistake was made. I thought maybe it was a relic from Flash, but AS3 also uses `false`. I'll fix this as well.

**Comment by gskinner** [comment]
I just pushed a branch:
https://github.com/CreateJS/EaselJS/tree/cancelableEventFixes

Please have a look and test as appropriate. I implemented your suggestions on `dispatchEvent`, fixed the `drawstart` and `tickstart` events to be cancelable, fixed the return value on `dispatchEvent`, and made a few documentation improvements.

**Comment by pittins** [comment]
Tossed out my local modifications to easeljs.0.8.0.combined, and threw easeljs.NEXT.combined into my project. I'm able to cancel both events, and am not noticing any aberrent behavior.

Gets my thumbs up. Thank you both for taking a look!

**Comment by gskinner** [comment]
Merged.
